ddtrace/appsec/processor.py

```diff
@@ -275,14 +329,28 @@ def on_span_finish(self, span):
             if body is not None:
                 data[_Addresses.SERVER_REQUEST_BODY] = body
 
+        if self._is_needed(_Addresses.HTTP_CLIENT_IP):
+            remote_ip = _context.get_item("http.request.remote_ip", span=span)
+            if remote_ip:
+                data[_Addresses.HTTP_CLIENT_IP] = remote_ip
+
         log.debug("[DDAS-001-00] Executing AppSec In-App WAF with parameters: %s", data)
         ddwaf_result = None
-        try:
-            ddwaf_result = self._ddwaf.run(data, self._waf_timeout)  # res is a serialized json
-        except OSError:
-            log.warning("Error executing Appsec In-App WAF: ", exc_info=True)
-        except Exception as e:
-            log.warning("Error executing Appsec In-App WAF: %s", repr(e))
+        total_runtime = 0.0
+        total_overall_runtime = 0.0
+        blocked_request = _context.get_item("http.request.blocked", span=span)
+        if not blocked_request:
```

Comment:
if the request wasn't originally blocked, we want to run the waf again to double check?

I assume this is in case additional information came up during the lifetime of the request that would make the request be blocked?


if a request is valid/good then we will always run the waf twice, once at the beginning of the request, and once at the end to see if it should be blocked ?


(this seems valid, just typing it out so confirm my understanding)

Reply:
Yup, the initial check is only for IP blocking with the early information (the IP), but we could block later once we have more information available which means 1/2 calls to the WAF and could be even more since for example the user blocking feature is a function that the user can call and must call the WAF. My gut sense is that once we've merged all the blocking features we can see some way to optimize all this since we'll know all the potential call points for the WAF.